Name: Carlos Constante

In [31]:
#importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.metrics import confusion_matrix
from sklearn.metrics import silhouette_score

from sklearn.preprocessing import MinMaxScaler

from sklearn.cluster import KMeans

from sklearn.datasets import make_blobs

%matplotlib inline
pd.set_option('display.max_columns', None)

In [32]:
#Loading dataframe
sal = pd.read_csv('NYC_Payroll_Data_Fiscal_Year.csv')
sal

Fiscal Year  Payroll Number                     Agency Name  \
0               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
1               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
2               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
3               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
4               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
...              ...             ...                             ...   
1469652         2019              56               POLICE DEPARTMENT   
1469653         2019              56               POLICE DEPARTMENT   
1469654         2019              56               POLICE DEPARTMENT   
1469655         2019              56               POLICE DEPARTMENT   
1469656         2019              56               POLICE DEPARTMENT   

            Last Name First Name Mid Init Agency Start Date  \
0             BEREZIN    MIKHAIL      NaN        08/10/2015   
1              GEAGER   VERONICA        M        09/12/2016   
2              RAMANI   SHRADDHA      NaN        02/22/2016   
3               ROTTA   JONATHAN        D        09/16/2013   
4           WILSON II     ROBERT        P        04/30/2018   
...               ...        ...      ...               ...   
1469652  LUNGAYAN JR.     HERMAN        V        05/28/2014   
1469653         HORNE   JONISHAU      NaN        02/20/2018   
1469654          HIRA      ABDUL        M        09/25/2017   
1469655         KARAI      WAHID        F        02/23/2016   
1469656       SHEHZAD   MOHAMMAD        A        12/22/2015   

        Work Location Borough                        Title Description  \
0                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
1                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
2                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
3                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
4                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
...                       ...                                      ...   
1469652             MANHATTAN  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY   
1469653              BROOKLYN         POLICE COMMUNICATIONS TECHNICIAN   
1469654                 BRONX  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY   
1469655              BROOKLYN         POLICE COMMUNICATIONS TECHNICIAN   
1469656                QUEENS  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY   

        Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                           ACTIVE      86005.0  per Annum         1820.0   
1                           ACTIVE      86005.0  per Annum         1820.0   
2                           ACTIVE      86005.0  per Annum         1820.0   
3                           ACTIVE      86005.0  per Annum         1820.0   
4                           ACTIVE      86005.0  per Annum         1820.0   
...                            ...          ...        ...            ...   
1469652                     ACTIVE      39449.0  per Annum         2040.0   
1469653                     ACTIVE      39453.0  per Annum         1785.0   
1469654                     ACTIVE      38986.0  per Annum         2040.0   
1469655                     ACTIVE      51700.0  per Annum         1785.0   
1469656                     ACTIVE      38986.0  per Annum         2040.0   

         Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0                  84698.21      0.00           0.00             0.00  
1                  84698.21      0.00           0.00             0.00  
2                  84698.21      0.00           0.00             0.00  
3                  84698.21      0.00           0.00             0.00  
4                  84698.21      0.00           0.00             0.00  
...                     ...       ...            ...              ...  
1469652            39032.61    306.50        9242.39          4182.91

In [33]:
#Check the dataframe columns' names
sal.columns

Index(['Fiscal Year', 'Payroll Number', 'Agency Name', 'Last Name',
       'First Name', 'Mid Init', 'Agency Start Date', 'Work Location Borough',
       'Title Description', 'Leave Status as of June 30', 'Base Salary',
       'Pay Basis', 'Regular Hours', 'Regular Gross Paid', 'OT Hours',
       'Total OT Paid', 'Total Other Pay'],
      dtype='object')

In [34]:
#check the dataframe information
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1469657 entries, 0 to 1469656
Data columns (total 17 columns):
 #   Column                      Non-Null Count    Dtype  
---  ------                      --------------    -----  
 0   Fiscal Year                 1469657 non-null  int64  
 1   Payroll Number              1469657 non-null  int64  
 2   Agency Name                 1469657 non-null  object 
 3   Last Name                   1468753 non-null  object 
 4   First Name                  1468742 non-null  object 
 5   Mid Init                    866071 non-null   object 
 6   Agency Start Date           1469616 non-null  object 
 7   Work Location Borough       1469651 non-null  object 
 8   Title Description           1469623 non-null  object 
 9   Leave Status as of June 30  1469657 non-null  object 
 10  Base Salary                 1469657 non-null  float64
 11  Pay Basis                   1469657 non-null  object 
 12  Regular Hours               1469657 non-null  float64
 1

In [35]:
#check for missing data in each numeric types column
sal.describe()

Fiscal Year  Payroll Number   Base Salary  Regular Hours  \
count  1.469657e+06    1.469657e+06  1.469657e+06   1.469657e+06   
mean   2.018996e+03    5.563574e+02  4.681239e+04   7.432658e+02   
std    8.130180e-01    3.090461e+02  4.388318e+04   9.178171e+02   
min    2.018000e+03    2.000000e+00  1.000000e+00  -2.352500e+02   
25%    2.018000e+03    3.000000e+02  3.318000e+01   0.000000e+00   
50%    2.019000e+03    7.420000e+02  4.497500e+04   0.000000e+00   
75%    2.020000e+03    7.470000e+02  8.529200e+04   1.820000e+03   
max    2.020000e+03    9.960000e+02  4.147070e+05   4.160000e+03   

       Regular Gross Paid      OT Hours  Total OT Paid  Total Other Pay  
count        1.469657e+06  1.469657e+06   1.469657e+06     1.469657e+06  
mean         4.964219e+04  7.081650e+01   3.772085e+03     3.205983e+03  
std          4.141916e+04  1.656371e+02   9.570726e+03     5.980867e+03  
min         -2.343844e+04 -2.100000e+01  -2.649388e+04    -1.548196e+05  
25%          7.388030e+03  0.000000e+00   0.000000e+00     0.000000e+00  
50%          4.529081e+04  0.000000e+00   0.000000e+00     1.650000e+01  
75%          8.418630e+04  1.475000e+01   6.280900e+02     4.129790e+03  
max          3.685975e+05  3.147000e+03   2.373897e+05     2.222470e+05

In [36]:
#Decribe the string types columns in the dataframe
sal.describe(include = ["O"])

Agency Name  Last Name First Name Mid Init  \
count                  1469657    1468753    1468742   866071   
unique                     155     119034      65155       35   
top     DEPT OF ED PEDAGOGICAL  RODRIGUEZ    MICHAEL        A   
freq                    285125       9397      22282   130531   

       Agency Start Date Work Location Borough     Title Description  \
count            1469616               1469651               1469623   
unique             12994                    18                  1570   
top           01/01/2010             MANHATTAN  TEACHER- PER SESSION   
freq               29345               1050354                233441   

       Leave Status as of June 30  Pay Basis  
count                     1469657    1469657  
unique                          1          4  
top                        ACTIVE  per Annum  
freq                      1469657     898331

In [37]:
#create a dataframe with only rows missing at least one value
sal.isna().any(axis = 1)

0           True
1          False
2           True
3          False
4          False
           ...  
1469652    False
1469653     True
1469654    False
1469655    False
1469656    False
Length: 1469657, dtype: bool

In [38]:
#Create a filter with all rows that contain at least one missing data
nan_filter = sal.isna().any(axis = 1)
sal[nan_filter]

Fiscal Year  Payroll Number                     Agency Name  \
0               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
2               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
6               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
9               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
13              2020              17  OFFICE OF EMERGENCY MANAGEMENT   
...              ...             ...                             ...   
1469646         2019              56               POLICE DEPARTMENT   
1469648         2019              56               POLICE DEPARTMENT   
1469649         2019              56               POLICE DEPARTMENT   
1469650         2019              56               POLICE DEPARTMENT   
1469653         2019              56               POLICE DEPARTMENT   

              Last Name    First Name Mid Init Agency Start Date  \
0               BEREZIN       MIKHAIL      NaN        08/10/2015   
2                RAMANI      SHRADDHA      NaN        02/22/2016   
6               VAZQUEZ      MARGARET      NaN        09/29/2008   
9         DE LOS SANTOS        JANIRA      NaN        06/05/2017   
13             WILLIAMS        LATOYA      NaN        02/16/2005   
...                 ...           ...      ...               ...   
1469646          MIKHNO          INNA      NaN        07/02/2018   
1469648           ISLAM  A B M KAMRUL      NaN        05/31/2017   
1469649  CASTRO VASQUEZ        ADISON      NaN        07/22/2016   
1469650          JALLIM        ALISEN      NaN        01/16/2018   
1469653           HORNE      JONISHAU      NaN        02/20/2018   

        Work Location Borough                        Title Description  \
0                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
2                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
6                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
9                    BROOKLYN        EMERGENCY PREPAREDNESS SPECIALIST   
13                   BROOKLYN        EMERGENCY PREPAREDNESS SPECIALIST   
...                       ...                                      ...   
1469646              BROOKLYN                           POLICE OFFICER   
1469648             MANHATTAN  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY   
1469649                 BRONX                    SCHOOL CROSSING GUARD   
1469650             MANHATTAN  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY   
1469653              BROOKLYN         POLICE COMMUNICATIONS TECHNICIAN   

        Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                           ACTIVE      86005.0  per Annum        1820.00   
2                           ACTIVE      86005.0  per Annum        1820.00   
6                           ACTIVE      94415.0  per Annum        1820.00   
9                           ACTIVE      67676.0  per Annum        1820.00   
13                          ACTIVE      73403.0  per Annum        1820.00   
...                            ...          ...        ...            ...   
1469646                     ACTIVE      42500.0  per Annum        2000.00   
1469648                     ACTIVE      38986.0  per Annum        2040.00   
1469649                     ACTIVE         15.0   per Hour         952.65   
1469650                     ACTIVE      38986.0  per Annum        2040.00   
1469653                     ACTIVE      39453.0  per Annum        1785.00   

         Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0                  84698.21      0.00           0.00             0.00  
2                  84698.21      0.00           0.00             0.00  
6                  84312.72      0.00           0.00             0.00  
9                  66647.77    348.50       16572.64           144.15  
13                 72287.70    217.75        9636.04           258.44  
...                     ...       ...            ...              ...  
1469646  

In [39]:
#create a dataframe with only rows missing at least one value in Mid Init column
missing_rows = sal['Mid Init'].isna()
missing_rows

0           True
1          False
2           True
3          False
4          False
           ...  
1469652    False
1469653     True
1469654    False
1469655    False
1469656    False
Name: Mid Init, Length: 1469657, dtype: bool

In [40]:
#Checking the dataframe with missing values in the Mid Init column
sal[missing_rows]

Fiscal Year  Payroll Number                     Agency Name  \
0               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
2               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
6               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
9               2020              17  OFFICE OF EMERGENCY MANAGEMENT   
13              2020              17  OFFICE OF EMERGENCY MANAGEMENT   
...              ...             ...                             ...   
1469646         2019              56               POLICE DEPARTMENT   
1469648         2019              56               POLICE DEPARTMENT   
1469649         2019              56               POLICE DEPARTMENT   
1469650         2019              56               POLICE DEPARTMENT   
1469653         2019              56               POLICE DEPARTMENT   

              Last Name    First Name Mid Init Agency Start Date  \
0               BEREZIN       MIKHAIL      NaN        08/10/2015   
2                RAMANI      SHRADDHA      NaN        02/22/2016   
6               VAZQUEZ      MARGARET      NaN        09/29/2008   
9         DE LOS SANTOS        JANIRA      NaN        06/05/2017   
13             WILLIAMS        LATOYA      NaN        02/16/2005   
...                 ...           ...      ...               ...   
1469646          MIKHNO          INNA      NaN        07/02/2018   
1469648           ISLAM  A B M KAMRUL      NaN        05/31/2017   
1469649  CASTRO VASQUEZ        ADISON      NaN        07/22/2016   
1469650          JALLIM        ALISEN      NaN        01/16/2018   
1469653           HORNE      JONISHAU      NaN        02/20/2018   

        Work Location Borough                        Title Description  \
0                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
2                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
6                    BROOKLYN           EMERGENCY PREPAREDNESS MANAGER   
9                    BROOKLYN        EMERGENCY PREPAREDNESS SPECIALIST   
13                   BROOKLYN        EMERGENCY PREPAREDNESS SPECIALIST   
...                       ...                                      ...   
1469646              BROOKLYN                           POLICE OFFICER   
1469648             MANHATTAN  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY   
1469649                 BRONX                    SCHOOL CROSSING GUARD   
1469650             MANHATTAN  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY   
1469653              BROOKLYN         POLICE COMMUNICATIONS TECHNICIAN   

        Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                           ACTIVE      86005.0  per Annum        1820.00   
2                           ACTIVE      86005.0  per Annum        1820.00   
6                           ACTIVE      94415.0  per Annum        1820.00   
9                           ACTIVE      67676.0  per Annum        1820.00   
13                          ACTIVE      73403.0  per Annum        1820.00   
...                            ...          ...        ...            ...   
1469646                     ACTIVE      42500.0  per Annum        2000.00   
1469648                     ACTIVE      38986.0  per Annum        2040.00   
1469649                     ACTIVE         15.0   per Hour         952.65   
1469650                     ACTIVE      38986.0  per Annum        2040.00   
1469653                     ACTIVE      39453.0  per Annum        1785.00   

         Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0                  84698.21      0.00           0.00             0.00  
2                  84698.21      0.00           0.00             0.00  
6                  84312.72      0.00           0.00             0.00  
9                  66647.77    348.50       16572.64           144.15  
13                 72287.70    217.75        9636.04           258.44  
...                     ...       ...            ...              ...  
1469646  

This exploratory analysis shows that in the column 'Mid Init' there are 603,586 rows with missing data out of the total rows with missing data 603,610. Therefore with can delete this column. Moreover, columns 'Leave Status as of June 30' & 'Payroll Number' are no needed for this analysis, therefore they will be deleted as well.

In [41]:
#Delete the column "Mid Init","Leave Status as of June 30" & 'Payroll Number'
del sal['Mid Init']
del sal['Leave Status as of June 30'] 
del sal['Payroll Number'] 
sal

Fiscal Year                     Agency Name     Last Name First Name  \
0               2020  OFFICE OF EMERGENCY MANAGEMENT       BEREZIN    MIKHAIL   
1               2020  OFFICE OF EMERGENCY MANAGEMENT        GEAGER   VERONICA   
2               2020  OFFICE OF EMERGENCY MANAGEMENT        RAMANI   SHRADDHA   
3               2020  OFFICE OF EMERGENCY MANAGEMENT         ROTTA   JONATHAN   
4               2020  OFFICE OF EMERGENCY MANAGEMENT     WILSON II     ROBERT   
...              ...                             ...           ...        ...   
1469652         2019               POLICE DEPARTMENT  LUNGAYAN JR.     HERMAN   
1469653         2019               POLICE DEPARTMENT         HORNE   JONISHAU   
1469654         2019               POLICE DEPARTMENT          HIRA      ABDUL   
1469655         2019               POLICE DEPARTMENT         KARAI      WAHID   
1469656         2019               POLICE DEPARTMENT       SHEHZAD   MOHAMMAD   

        Agency Start Date Work Location Borough  \
0              08/10/2015              BROOKLYN   
1              09/12/2016              BROOKLYN   
2              02/22/2016              BROOKLYN   
3              09/16/2013              BROOKLYN   
4              04/30/2018              BROOKLYN   
...                   ...                   ...   
1469652        05/28/2014             MANHATTAN   
1469653        02/20/2018              BROOKLYN   
1469654        09/25/2017                 BRONX   
1469655        02/23/2016              BROOKLYN   
1469656        12/22/2015                QUEENS   

                               Title Description  Base Salary  Pay Basis  \
0                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
1                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
2                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
3                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
4                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
...                                          ...          ...        ...   
1469652  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      39449.0  per Annum   
1469653         POLICE COMMUNICATIONS TECHNICIAN      39453.0  per Annum   
1469654  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   
1469655         POLICE COMMUNICATIONS TECHNICIAN      51700.0  per Annum   
1469656  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   

         Regular Hours  Regular Gross Paid  OT Hours  Total OT Paid  \
0               1820.0            84698.21      0.00           0.00   
1               1820.0            84698.21      0.00           0.00   
2               1820.0            84698.21      0.00           0.00   
3               1820.0            84698.21      0.00           0.00   
4               1820.0            84698.21      0.00           0.00   
...                ...                 ...       ...            ...   
1469652         2040.0            39032.61    306.50        9242.39   
1469653         1785.0            38539.44    245.55        7163.58   
1469654         2040.0            38763.31    269.00        7979.68   
1469655         1785.0            49625.00    255.32        9214.59   
1469656         2040.0            39276.28    637.50       19430.30   

         Total Other Pay  
0                   0.00  
1                   0.00  
2                   0.00  
3                   0.00  
4                   0.00  
...                  ...  
1469652          4182.91  
1469653          3680.55  
1469654          1031.37  
1469655          8692.69  
1469656          2090.16  

[1469657 rows x 14 columns]

In [43]:
#get info about the 14 columns left
sal.describe(include = 'all')

Fiscal Year             Agency Name  Last Name First Name  \
count   1.469657e+06                 1469657    1468753    1468742   
unique           NaN                     155     119034      65155   
top              NaN  DEPT OF ED PEDAGOGICAL  RODRIGUEZ    MICHAEL   
freq             NaN                  285125       9397      22282   
mean    2.018996e+03                     NaN        NaN        NaN   
std     8.130180e-01                     NaN        NaN        NaN   
min     2.018000e+03                     NaN        NaN        NaN   
25%     2.018000e+03                     NaN        NaN        NaN   
50%     2.019000e+03                     NaN        NaN        NaN   
75%     2.020000e+03                     NaN        NaN        NaN   
max     2.020000e+03                     NaN        NaN        NaN   

       Agency Start Date Work Location Borough     Title Description  \
count            1469616               1469651               1469623   
unique             12994                    18                  1570   
top           01/01/2010             MANHATTAN  TEACHER- PER SESSION   
freq               29345               1050354                233441   
mean                 NaN                   NaN                   NaN   
std                  NaN                   NaN                   NaN   
min                  NaN                   NaN                   NaN   
25%                  NaN                   NaN                   NaN   
50%                  NaN                   NaN                   NaN   
75%                  NaN                   NaN                   NaN   
max                  NaN                   NaN                   NaN   

         Base Salary  Pay Basis  Regular Hours  Regular Gross Paid  \
count   1.469657e+06    1469657   1.469657e+06        1.469657e+06   
unique           NaN          4            NaN                 NaN   
top              NaN  per Annum            NaN                 NaN   
freq             NaN     898331            NaN                 NaN   
mean    4.681239e+04        NaN   7.432658e+02        4.964219e+04   
std     4.388318e+04        NaN   9.178171e+02        4.141916e+04   
min     1.000000e+00        NaN  -2.352500e+02       -2.343844e+04   
25%     3.318000e+01        NaN   0.000000e+00        7.388030e+03   
50%     4.497500e+04        NaN   0.000000e+00        4.529081e+04   
75%     8.529200e+04        NaN   1.820000e+03        8.418630e+04   
max     4.147070e+05        NaN   4.160000e+03        3.685975e+05   

            OT Hours  Total OT Paid  Total Other Pay  
count   1.469657e+06   1.469657e+06     1.469657e+06  
unique           NaN            NaN              NaN  
top              NaN            NaN              NaN  
freq             NaN            NaN              NaN  
mean    7.081650e+01   3.772085e+03     3.205983e+03  
std     1.656371e+02   9.570726e+03     5.980867e+03  
min    -2.100000e+01  -2.649388e+04    -1.548196e+05  
25%     0.000000e+00   0.000000e+00     0.000000e+00  
50%     0.000000e+00   0.000000e+00     1.650000e+01  
75%     1.475000e+01   6.280900e+02     4.129790e+03  
max     3.147000e+03   2.373897e+05     2.222470e+05

In [44]:
#Creating a new filter for missing values for the 14 columns left
nan_filter = sal.isna().any(axis = 1)
sal[nan_filter]

Fiscal Year                     Agency Name Last Name First Name  \
114216          2020  BOARD OF ELECTION POLL WORKERS    FEUCHT        NaN   
128167          2020  BOARD OF ELECTION POLL WORKERS       NaN       YOUN   
140336          2020          DEPT OF ED PEDAGOGICAL   COLEMAN     KAWANA   
235224          2020  BOARD OF ELECTION POLL WORKERS       NaN     DANIEL   
235225          2020  BOARD OF ELECTION POLL WORKERS       NaN     MIJEAN   
...              ...                             ...       ...        ...   
1404519         2019  DISTRICT ATTORNEY-SPECIAL NARC       NaN        NaN   
1404520         2019  DISTRICT ATTORNEY-SPECIAL NARC       NaN        NaN   
1404540         2019  DISTRICT ATTORNEY-SPECIAL NARC       NaN        NaN   
1404564         2019  DISTRICT ATTORNEY-SPECIAL NARC       NaN        NaN   
1404566         2019  DISTRICT ATTORNEY-SPECIAL NARC       NaN        NaN   

        Agency Start Date Work Location Borough  \
114216         01/01/2019             MANHATTAN   
128167         01/01/2018             MANHATTAN   
140336         07/01/1982                   NaN   
235224         01/01/2019             MANHATTAN   
235225         01/01/2019             MANHATTAN   
...                   ...                   ...   
1404519        06/11/2018             MANHATTAN   
1404520        06/11/2018             MANHATTAN   
1404540        09/04/2018             MANHATTAN   
1404564        04/01/2019             MANHATTAN   
1404566        04/01/2019             MANHATTAN   

                                         Title Description  Base Salary  \
114216                                     ELECTION WORKER          1.0   
128167                                     ELECTION WORKER          1.0   
140336                           TEACHER SPECIAL EDUCATION     124909.0   
235224                                     ELECTION WORKER          1.0   
235225                                     ELECTION WORKER          1.0   
...                                                    ...          ...   
1404519  SENIOR RACKETS INVESTIGATOR - START >4-24-08 N...      63049.0   
1404520  SENIOR RACKETS INVESTIGATOR - START >4-24-08 N...      63012.0   
1404540  SENIOR RACKETS INVESTIGATOR - START >4-24-08 N...      65000.0   
1404564  SENIOR RACKETS INVESTIGATOR - START >4-24-08 N...      65000.0   
1404566       RACKETS INVESTIGATOR - START >4-24-08 NO ABC      60000.0   

         Pay Basis  Regular Hours  Regular Gross Paid  OT Hours  \
114216    per Hour            0.0             3995.00       0.0   
128167    per Hour            0.0              525.00       0.0   
140336   per Annum            0.0           122242.80       0.0   
235224    per Hour            0.0              350.00       0.0   
235225    per Hour            0.0              350.00       0.0   
...            ...            ...                 ...       ...   
1404519  per Annum         2040.0            64442.53       0.0   
1404520  per Annum         2040.0            64407.84       0.0   
1404540  per Annum         1672.0            53106.98       0.0   
1404564  per Annum          480.0            14958.90       0.0   
1404566  per Annum          480.0            13808.22       7.0   

         Total OT Paid  Total Other Pay  
114216            0.00             0.00  
128167            0.00             0.00  
140336            0.00         26460.23  
235224            0.00             0.00  
235225            0.00             0.00  
...                ...              ...  
1404519           0.00          3626.43  
1404520           0.00          3626.43  
1404540           0.00            16.50  
1404564           0.00           251.38  
1404566         301.73             0.00  

[938 rows x 14 columns]

There were 938 rows with missing data such as 'First name', 'Lastname', 'Agency Start Date', 'Work Location Borough', and 'Title Description'. Those rows will be dropped. This could be possibly human errors while entering the data.

In [53]:
sal[nan_filter].describe(include = 'all')

Fiscal Year                  Agency Name Last Name First Name  \
count    938.000000                          938        34         23   
unique          NaN                           15        13         10   
top             NaN  DISTRICT ATTORNEY-MANHATTAN        XU     YUNJIN   
freq            NaN                          251         6          3   
mean    2018.989339                          NaN       NaN        NaN   
std        0.813371                          NaN       NaN        NaN   
min     2018.000000                          NaN       NaN        NaN   
25%     2018.000000                          NaN       NaN        NaN   
50%     2019.000000                          NaN       NaN        NaN   
75%     2020.000000                          NaN       NaN        NaN   
max     2020.000000                          NaN       NaN        NaN   

       Agency Start Date Work Location Borough  \
count                897                   932   
unique               288                     5   
top           01/09/2017             MANHATTAN   
freq                  10                   331   
mean                 NaN                   NaN   
std                  NaN                   NaN   
min                  NaN                   NaN   
25%                  NaN                   NaN   
50%                  NaN                   NaN   
75%                  NaN                   NaN   
max                  NaN                   NaN   

                                        Title Description    Base Salary  \
count                                                 904     938.000000   
unique                                                 22            NaN   
top     SENIOR RACKETS INVESTIGATOR - START >4-24-08 N...            NaN   
freq                                                  407            NaN   
mean                                                  NaN   74848.996962   
std                                                   NaN   22249.579315   
min                                                   NaN       1.000000   
25%                                                   NaN   62751.500000   
50%                                                   NaN   70000.000000   
75%                                                   NaN   84818.000000   
max                                                   NaN  188568.000000   

        Pay Basis  Regular Hours  Regular Gross Paid     OT Hours  \
count         938     938.000000          938.000000   938.000000   
unique          2            NaN                 NaN          NaN   
top     per Annum            NaN                 NaN          NaN   
freq          920            NaN                 NaN          NaN   
mean          NaN    1910.463731        73130.904051   164.050640   
std           NaN     426.280148        24170.137307   214.686179   
min           NaN       0.000000          250.000000     0.000000   
25%           NaN    2040.000000        63400.160000     3.000000   
50%           NaN    2045.720000        70527.215000    74.250000   
75%           NaN    2080.000000        83949.750000   262.125000   
max           NaN    2085.720000       186592.220000  1711.000000   

        Total OT Paid  Total Other Pay  
count      938.000000       938.000000  
unique            NaN              NaN  
top               NaN              NaN  
freq              NaN              NaN  
mean      8657.375299      5624.295085  
std      11454.590988      5690.216186  
min          0.000000         0.000000  
25%        181.660000       855.640000  
50%       3639.155000      3915.440000  
75%      13537.505000      9049.675000  
max      80693.170000     49263.740000

In [56]:
#Removing rows with missing values and creating a new dataframe
sal2 = sal.dropna()
sal2

Fiscal Year                     Agency Name     Last Name First Name  \
0               2020  OFFICE OF EMERGENCY MANAGEMENT       BEREZIN    MIKHAIL   
1               2020  OFFICE OF EMERGENCY MANAGEMENT        GEAGER   VERONICA   
2               2020  OFFICE OF EMERGENCY MANAGEMENT        RAMANI   SHRADDHA   
3               2020  OFFICE OF EMERGENCY MANAGEMENT         ROTTA   JONATHAN   
4               2020  OFFICE OF EMERGENCY MANAGEMENT     WILSON II     ROBERT   
...              ...                             ...           ...        ...   
1469652         2019               POLICE DEPARTMENT  LUNGAYAN JR.     HERMAN   
1469653         2019               POLICE DEPARTMENT         HORNE   JONISHAU   
1469654         2019               POLICE DEPARTMENT          HIRA      ABDUL   
1469655         2019               POLICE DEPARTMENT         KARAI      WAHID   
1469656         2019               POLICE DEPARTMENT       SHEHZAD   MOHAMMAD   

        Agency Start Date Work Location Borough  \
0              08/10/2015              BROOKLYN   
1              09/12/2016              BROOKLYN   
2              02/22/2016              BROOKLYN   
3              09/16/2013              BROOKLYN   
4              04/30/2018              BROOKLYN   
...                   ...                   ...   
1469652        05/28/2014             MANHATTAN   
1469653        02/20/2018              BROOKLYN   
1469654        09/25/2017                 BRONX   
1469655        02/23/2016              BROOKLYN   
1469656        12/22/2015                QUEENS   

                               Title Description  Base Salary  Pay Basis  \
0                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
1                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
2                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
3                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
4                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
...                                          ...          ...        ...   
1469652  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      39449.0  per Annum   
1469653         POLICE COMMUNICATIONS TECHNICIAN      39453.0  per Annum   
1469654  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   
1469655         POLICE COMMUNICATIONS TECHNICIAN      51700.0  per Annum   
1469656  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   

         Regular Hours  Regular Gross Paid  OT Hours  Total OT Paid  \
0               1820.0            84698.21      0.00           0.00   
1               1820.0            84698.21      0.00           0.00   
2               1820.0            84698.21      0.00           0.00   
3               1820.0            84698.21      0.00           0.00   
4               1820.0            84698.21      0.00           0.00   
...                ...                 ...       ...            ...   
1469652         2040.0            39032.61    306.50        9242.39   
1469653         1785.0            38539.44    245.55        7163.58   
1469654         2040.0            38763.31    269.00        7979.68   
1469655         1785.0            49625.00    255.32        9214.59   
1469656         2040.0            39276.28    637.50       19430.30   

         Total Other Pay  
0                   0.00  
1                   0.00  
2                   0.00  
3                   0.00  
4                   0.00  
...                  ...  
1469652          4182.91  
1469653          3680.55  
1469654          1031.37  
1469655          8692.69  
1469656          2090.16  

[1468719 rows x 14 columns]

In [49]:
#Describing all columns in new dataframe
sal2.describe(include = 'all')

Fiscal Year             Agency Name  Last Name First Name  \
count   1.468719e+06                 1468719    1468719    1468719   
unique           NaN                     155     119033      65153   
top              NaN  DEPT OF ED PEDAGOGICAL  RODRIGUEZ    MICHAEL   
freq             NaN                  285121       9397      22282   
mean    2.018996e+03                     NaN        NaN        NaN   
std     8.130180e-01                     NaN        NaN        NaN   
min     2.018000e+03                     NaN        NaN        NaN   
25%     2.018000e+03                     NaN        NaN        NaN   
50%     2.019000e+03                     NaN        NaN        NaN   
75%     2.020000e+03                     NaN        NaN        NaN   
max     2.020000e+03                     NaN        NaN        NaN   

       Agency Start Date Work Location Borough     Title Description  \
count            1468719               1468719               1468719   
unique             12993                    18                  1563   
top           01/01/2010             MANHATTAN  TEACHER- PER SESSION   
freq               29345               1050023                233441   
mean                 NaN                   NaN                   NaN   
std                  NaN                   NaN                   NaN   
min                  NaN                   NaN                   NaN   
25%                  NaN                   NaN                   NaN   
50%                  NaN                   NaN                   NaN   
75%                  NaN                   NaN                   NaN   
max                  NaN                   NaN                   NaN   

         Base Salary  Pay Basis  Regular Hours  Regular Gross Paid  \
count   1.468719e+06    1468719   1.468719e+06        1.468719e+06   
unique           NaN          4            NaN                 NaN   
top              NaN  per Annum            NaN                 NaN   
freq             NaN     897411            NaN                 NaN   
mean    4.679449e+04        NaN   7.425204e+02        4.962718e+04   
std     4.388787e+04        NaN   9.175727e+02        4.142363e+04   
min     1.000000e+00        NaN  -2.352500e+02       -2.343844e+04   
25%     3.318000e+01        NaN   0.000000e+00        7.368430e+03   
50%     4.497500e+04        NaN   0.000000e+00        4.526054e+04   
75%     8.529200e+04        NaN   1.820000e+03        8.418765e+04   
max     4.147070e+05        NaN   4.160000e+03        3.685975e+05   

            OT Hours  Total OT Paid  Total Other Pay  
count   1.468719e+06   1.468719e+06     1.468719e+06  
unique           NaN            NaN              NaN  
top              NaN            NaN              NaN  
freq             NaN            NaN              NaN  
mean    7.075696e+01   3.768965e+03     3.204439e+03  
std     1.655844e+02   9.568612e+03     5.980738e+03  
min    -2.100000e+01  -2.649388e+04    -1.548196e+05  
25%     0.000000e+00   0.000000e+00     0.000000e+00  
50%     0.000000e+00   0.000000e+00     1.650000e+01  
75%     1.450000e+01   6.197500e+02     4.123980e+03  
max     3.147000e+03   2.373897e+05     2.222470e+05

In [52]:
#check the job titles in this dataframe
sal2['Work Location Borough'].value_counts()

MANHATTAN        1050023
QUEENS            164518
BROOKLYN          135857
BRONX              74745
RICHMOND           20552
OTHER              19471
WESTCHESTER         1550
ULSTER               936
SULLIVAN             362
DELAWARE             243
PUTNAM               114
NASSAU               101
SCHOHARIE             84
DUTCHESS              68
ALBANY                43
GREENE                30
WASHINGTON DC         19
ORANGE                 3
Name: Work Location Borough, dtype: int64

In [60]:
# Filtering the dataframe 'Work Location' with Manhattan, Queens, Brooklyn and Bronx
sal3 = sal2[sal2['Work Location Borough'].isin(['MANHATTAN','QUEENS','BROOKLYN','BRONX'])]
sal3

Fiscal Year                     Agency Name     Last Name First Name  \
0               2020  OFFICE OF EMERGENCY MANAGEMENT       BEREZIN    MIKHAIL   
1               2020  OFFICE OF EMERGENCY MANAGEMENT        GEAGER   VERONICA   
2               2020  OFFICE OF EMERGENCY MANAGEMENT        RAMANI   SHRADDHA   
3               2020  OFFICE OF EMERGENCY MANAGEMENT         ROTTA   JONATHAN   
4               2020  OFFICE OF EMERGENCY MANAGEMENT     WILSON II     ROBERT   
...              ...                             ...           ...        ...   
1469652         2019               POLICE DEPARTMENT  LUNGAYAN JR.     HERMAN   
1469653         2019               POLICE DEPARTMENT         HORNE   JONISHAU   
1469654         2019               POLICE DEPARTMENT          HIRA      ABDUL   
1469655         2019               POLICE DEPARTMENT         KARAI      WAHID   
1469656         2019               POLICE DEPARTMENT       SHEHZAD   MOHAMMAD   

        Agency Start Date Work Location Borough  \
0              08/10/2015              BROOKLYN   
1              09/12/2016              BROOKLYN   
2              02/22/2016              BROOKLYN   
3              09/16/2013              BROOKLYN   
4              04/30/2018              BROOKLYN   
...                   ...                   ...   
1469652        05/28/2014             MANHATTAN   
1469653        02/20/2018              BROOKLYN   
1469654        09/25/2017                 BRONX   
1469655        02/23/2016              BROOKLYN   
1469656        12/22/2015                QUEENS   

                               Title Description  Base Salary  Pay Basis  \
0                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
1                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
2                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
3                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
4                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
...                                          ...          ...        ...   
1469652  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      39449.0  per Annum   
1469653         POLICE COMMUNICATIONS TECHNICIAN      39453.0  per Annum   
1469654  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   
1469655         POLICE COMMUNICATIONS TECHNICIAN      51700.0  per Annum   
1469656  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   

         Regular Hours  Regular Gross Paid  OT Hours  Total OT Paid  \
0               1820.0            84698.21      0.00           0.00   
1               1820.0            84698.21      0.00           0.00   
2               1820.0            84698.21      0.00           0.00   
3               1820.0            84698.21      0.00           0.00   
4               1820.0            84698.21      0.00           0.00   
...                ...                 ...       ...            ...   
1469652         2040.0            39032.61    306.50        9242.39   
1469653         1785.0            38539.44    245.55        7163.58   
1469654         2040.0            38763.31    269.00        7979.68   
1469655         1785.0            49625.00    255.32        9214.59   
1469656         2040.0            39276.28    637.50       19430.30   

         Total Other Pay  
0                   0.00  
1                   0.00  
2                   0.00  
3                   0.00  
4                   0.00  
...                  ...  
1469652          4182.91  
1469653          3680.55  
1469654          1031.37  
1469655          8692.69  
1469656          2090.16  

[1425143 rows x 14 columns]

In [61]:
sal3.describe(include = 'all')

Fiscal Year             Agency Name  Last Name First Name  \
count   1.425143e+06                 1425143    1425143    1425143   
unique           NaN                     149     117235      64700   
top              NaN  DEPT OF ED PEDAGOGICAL  RODRIGUEZ    MICHAEL   
freq             NaN                  265655       9236      21079   
mean    2.018998e+03                     NaN        NaN        NaN   
std     8.129267e-01                     NaN        NaN        NaN   
min     2.018000e+03                     NaN        NaN        NaN   
25%     2.018000e+03                     NaN        NaN        NaN   
50%     2.019000e+03                     NaN        NaN        NaN   
75%     2.020000e+03                     NaN        NaN        NaN   
max     2.020000e+03                     NaN        NaN        NaN   

       Agency Start Date Work Location Borough     Title Description  \
count            1425143               1425143               1425143   
unique             12921                     4                  1537   
top           01/01/2010             MANHATTAN  TEACHER- PER SESSION   
freq               29345               1050023                233441   
mean                 NaN                   NaN                   NaN   
std                  NaN                   NaN                   NaN   
min                  NaN                   NaN                   NaN   
25%                  NaN                   NaN                   NaN   
50%                  NaN                   NaN                   NaN   
75%                  NaN                   NaN                   NaN   
max                  NaN                   NaN                   NaN   

         Base Salary  Pay Basis  Regular Hours  Regular Gross Paid  \
count   1.425143e+06    1425143   1.425143e+06        1.425143e+06   
unique           NaN          4            NaN                 NaN   
top              NaN  per Annum            NaN                 NaN   
freq             NaN     857981            NaN                 NaN   
mean    4.554316e+04        NaN   7.347325e+02        4.833214e+04   
std     4.332212e+04        NaN   9.151047e+02        4.077300e+04   
min     1.000000e+00        NaN  -2.352500e+02       -2.343844e+04   
25%     3.318000e+01        NaN   0.000000e+00        6.671615e+03   
50%     4.405200e+04        NaN   0.000000e+00        4.421395e+04   
75%     8.401250e+04        NaN   1.820000e+03        8.283452e+04   
max     4.147070e+05        NaN   4.160000e+03        3.685975e+05   

            OT Hours  Total OT Paid  Total Other Pay  
count   1.425143e+06   1.425143e+06     1.425143e+06  
unique           NaN            NaN              NaN  
top              NaN            NaN              NaN  
freq             NaN            NaN              NaN  
mean    6.953900e+01   3.691300e+03     3.123824e+03  
std     1.644188e+02   9.469556e+03     5.896730e+03  
min    -2.100000e+01  -2.649388e+04    -1.548196e+05  
25%     0.000000e+00   0.000000e+00     0.000000e+00  
50%     0.000000e+00   0.000000e+00     1.257000e+01  
75%     1.125000e+01   5.083850e+02     3.983990e+03  
max     3.147000e+03   2.150228e+05     2.222470e+05

In [62]:
#check the value counts in column 'Pay Basis'
sal3['Pay Basis'].value_counts()

per Annum          857981
per Day            328280
per Hour           233439
Prorated Annual      5443
Name: Pay Basis, dtype: int64

In [64]:
#Obtaining dataframe columns' name 
sal3.columns

Index(['Fiscal Year', 'Agency Name', 'Last Name', 'First Name',
       'Agency Start Date', 'Work Location Borough', 'Title Description',
       'Base Salary', 'Pay Basis', 'Regular Hours', 'Regular Gross Paid',
       'OT Hours', 'Total OT Paid', 'Total Other Pay'],
      dtype='object')

In [68]:
#Creating a new dataframe containing only rows with equal or greater than 0
sal_nyc = sal3[(sal3['Regular Hours'] >= 0) & \
              (sal3['Regular Gross Paid'] >= 0) & \
              (sal3['OT Hours'] >= 0) & \
              (sal3['Total OT Paid'] >= 0) & \
              (sal3['Total Other Pay'] >= 0) ]
sal_nyc

Fiscal Year                     Agency Name     Last Name First Name  \
0               2020  OFFICE OF EMERGENCY MANAGEMENT       BEREZIN    MIKHAIL   
1               2020  OFFICE OF EMERGENCY MANAGEMENT        GEAGER   VERONICA   
2               2020  OFFICE OF EMERGENCY MANAGEMENT        RAMANI   SHRADDHA   
3               2020  OFFICE OF EMERGENCY MANAGEMENT         ROTTA   JONATHAN   
4               2020  OFFICE OF EMERGENCY MANAGEMENT     WILSON II     ROBERT   
...              ...                             ...           ...        ...   
1469652         2019               POLICE DEPARTMENT  LUNGAYAN JR.     HERMAN   
1469653         2019               POLICE DEPARTMENT         HORNE   JONISHAU   
1469654         2019               POLICE DEPARTMENT          HIRA      ABDUL   
1469655         2019               POLICE DEPARTMENT         KARAI      WAHID   
1469656         2019               POLICE DEPARTMENT       SHEHZAD   MOHAMMAD   

        Agency Start Date Work Location Borough  \
0              08/10/2015              BROOKLYN   
1              09/12/2016              BROOKLYN   
2              02/22/2016              BROOKLYN   
3              09/16/2013              BROOKLYN   
4              04/30/2018              BROOKLYN   
...                   ...                   ...   
1469652        05/28/2014             MANHATTAN   
1469653        02/20/2018              BROOKLYN   
1469654        09/25/2017                 BRONX   
1469655        02/23/2016              BROOKLYN   
1469656        12/22/2015                QUEENS   

                               Title Description  Base Salary  Pay Basis  \
0                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
1                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
2                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
3                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
4                 EMERGENCY PREPAREDNESS MANAGER      86005.0  per Annum   
...                                          ...          ...        ...   
1469652  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      39449.0  per Annum   
1469653         POLICE COMMUNICATIONS TECHNICIAN      39453.0  per Annum   
1469654  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   
1469655         POLICE COMMUNICATIONS TECHNICIAN      51700.0  per Annum   
1469656  TRAFFIC ENFORCEMENT AGENT AL 1 & 2 ONLY      38986.0  per Annum   

         Regular Hours  Regular Gross Paid  OT Hours  Total OT Paid  \
0               1820.0            84698.21      0.00           0.00   
1               1820.0            84698.21      0.00           0.00   
2               1820.0            84698.21      0.00           0.00   
3               1820.0            84698.21      0.00           0.00   
4               1820.0            84698.21      0.00           0.00   
...                ...                 ...       ...            ...   
1469652         2040.0            39032.61    306.50        9242.39   
1469653         1785.0            38539.44    245.55        7163.58   
1469654         2040.0            38763.31    269.00        7979.68   
1469655         1785.0            49625.00    255.32        9214.59   
1469656         2040.0            39276.28    637.50       19430.30   

         Total Other Pay  
0                   0.00  
1                   0.00  
2                   0.00  
3                   0.00  
4                   0.00  
...                  ...  
1469652          4182.91  
1469653          3680.55  
1469654          1031.37  
1469655          8692.69  
1469656          2090.16  

[1415277 rows x 14 columns]

In [69]:
#Describing all columns in the new dataframe nyc
sal_nyc.describe(include = 'all')

Fiscal Year             Agency Name  Last Name First Name  \
count   1.415277e+06                 1415277    1415277    1415277   
unique           NaN                     149     117144      64649   
top              NaN  DEPT OF ED PEDAGOGICAL  RODRIGUEZ    MICHAEL   
freq             NaN                  262378       9160      20883   
mean    2.018995e+03                     NaN        NaN        NaN   
std     8.126307e-01                     NaN        NaN        NaN   
min     2.018000e+03                     NaN        NaN        NaN   
25%     2.018000e+03                     NaN        NaN        NaN   
50%     2.019000e+03                     NaN        NaN        NaN   
75%     2.020000e+03                     NaN        NaN        NaN   
max     2.020000e+03                     NaN        NaN        NaN   

       Agency Start Date Work Location Borough     Title Description  \
count            1415277               1415277               1415277   
unique             12919                     4                  1536   
top           01/01/2010             MANHATTAN  TEACHER- PER SESSION   
freq               29330               1043422                233365   
mean                 NaN                   NaN                   NaN   
std                  NaN                   NaN                   NaN   
min                  NaN                   NaN                   NaN   
25%                  NaN                   NaN                   NaN   
50%                  NaN                   NaN                   NaN   
75%                  NaN                   NaN                   NaN   
max                  NaN                   NaN                   NaN   

         Base Salary  Pay Basis  Regular Hours  Regular Gross Paid  \
count   1.415277e+06    1415277   1.415277e+06        1.415277e+06   
unique           NaN          4            NaN                 NaN   
top              NaN  per Annum            NaN                 NaN   
freq             NaN     848880            NaN                 NaN   
mean    4.540299e+04        NaN   7.337083e+02        4.821749e+04   
std     4.335722e+04        NaN   9.146393e+02        4.080839e+04   
min     1.000000e+00        NaN   0.000000e+00        0.000000e+00   
25%     3.318000e+01        NaN   0.000000e+00        6.523820e+03   
50%     4.394100e+04        NaN   0.000000e+00        4.412998e+04   
75%     8.392400e+04        NaN   1.820000e+03        8.273082e+04   
max     4.147070e+05        NaN   4.160000e+03        3.625225e+05   

            OT Hours  Total OT Paid  Total Other Pay  
count   1.415277e+06   1.415277e+06     1.415277e+06  
unique           NaN            NaN              NaN  
top              NaN            NaN              NaN  
freq             NaN            NaN              NaN  
mean    6.957002e+01   3.690437e+03     3.183051e+03  
std     1.645965e+02   9.474676e+03     5.785830e+03  
min     0.000000e+00   0.000000e+00     0.000000e+00  
25%     0.000000e+00   0.000000e+00     0.000000e+00  
50%     0.000000e+00   0.000000e+00     1.562000e+01  
75%     1.100000e+01   4.984300e+02     4.006020e+03  
max     3.147000e+03   2.150228e+05     2.222470e+05

In [72]:
len(sal3) - len(sal_nyc)

9866

In [71]:
len(sal_nyc)

1415277

We removed 9,866 rows. Compared to the total number of 1,415,277 this is a small number. Moreover, it improves the quality of the data by excluding negative values while not changing the distributions of the other columns very much due to the small amount of points were removed.

In [76]:
sal_nyc['Agency Name'].value_counts().head(25)

DEPT OF ED PEDAGOGICAL            262378
DEPT OF ED PER SESSION TEACHER    240341
POLICE DEPARTMENT                 155773
BOARD OF ELECTION POLL WORKERS    100757
DEPT OF ED PARA PROFESSIONALS      98918
DEPT OF ED HRLY SUPPORT STAFF      55633
FIRE DEPARTMENT                    47782
DEPARTMENT OF EDUCATION ADMIN      41257
HRA/DEPT OF SOCIAL SERVICES        37273
DEPT OF ED PER DIEM TEACHERS       35172
DEPARTMENT OF CORRECTION           34468
NYC HOUSING AUTHORITY              31763
DEPARTMENT OF SANITATION           27693
DEPT OF PARKS & RECREATION         26123
DEPT OF HEALTH/MENTAL HYGIENE      20721
ADMIN FOR CHILDREN'S SVCS          19987
DEPARTMENT OF TRANSPORTATION       14504
DEPT OF ENVIRONMENT PROTECTION     14183
COMMUNITY COLLEGE (LAGUARDIA)      10740
COMMUNITY COLLEGE (MANHATTAN)      10660
COMMUNITY COLLEGE (KINGSBORO)       7833
DEPT OF CITYWIDE ADMIN SVCS         7134
HOUSING PRESERVATION & DVLPMNT      7007
COMMUNITY COLLEGE (QUEENSBORO)      6961
DEPT. OF HOMELES

In [77]:
sal_nyc['Title Description'].value_counts().head(25)

TEACHER- PER SESSION                233365
TEACHER                             146393
ELECTION WORKER                      99683
ANNUAL ED PARA                       74045
TEACHER SPECIAL EDUCATION            67304
POLICE OFFICER                       66998
TEACHER-GENERAL ED                   32609
CORRECTION OFFICER                   26104
SUBSTITUTE ED PARA                   24114
FIREFIGHTER                          22336
F/T SCHOOL AIDE                      21299
SANITATION WORKER                    17180
COMMUNITY ASSOCIATE                  14775
SCHOOL SAFETY AGENT                  14178
COMMUNITY COORDINATOR                12485
COLLEGE ASSISTANT                    11690
CLERICAL ASSOCIATE                   11546
SERGEANT-                            11382
F/T SCHOOL LUNCH HELPER              10544
CARETAKER                             9459
P.O. DA DET GR3                       9426
ASSISTANT PRINCIPAL                   8998
GUIDANCE COUNSELOR                    8381
LIEUTENANT 

There are 149 agencies and 1,536 title jobs in the dataframe